lib/api/bucketPutTagging.js

```diff
@@ -0,0 +1,104 @@
+
+const collectCorsHeaders = require('../utilities/collectCorsHeaders');
+const { metadataValidateBucket } = require('../metadata/metadataUtils');
+const metadata = require('../metadata/wrapper');
+
+const tagKeyNotUnique = 'Cannot provide multiple Tags with the same key';
+const tagKeyInvalid = 'The TagKey you have provided is invalid';
+const tagValueInvalid = 'The TagValue you have provided is invalid';
+
+/**
+ * Format of xml request:
+
+ <Tagging xmlns="http://s3.amazonaws.com/doc/2006-03-01/">
+ <TagSet>
+ <Tag>
+ <Key>string</Key>
+ <Value>string</Value>
+ </Tag>
+ </TagSet>
+ </Tagging>
+ */
+
+function _parseXML(request, log, cb, parser) {
+    if (request.post === '') {
+        log.debug('request xml is missing');
+        return cb(errors.MalformedXML);
+    }
+    return parser.parseString(request.post, (err, result) => {
+        if (err) {
+            log.debug('request xml is malformed');
+            return cb(errors.MalformedXML);
+        }
+        return process.nextTick(() => cb(null, result));
+    });
+}
+
+function checkTags(tagSet, bucket, log, cb) {
+    if (tagSet === undefined) {
+        return cb(null, { tags: [], bucket });
+    }
+    const tags = Array.isArray(tagSet) ? tagSet : [tagSet];
+    const unique = [...new Set(tags.map(item => item.Key))];
+    let errMsg = null;
+    if (unique.length !== tags.length) {
+        errMsg = tagKeyNotUnique;
+    }
+    tags.forEach(item => {
+        if (item.Key.length > 128) {
+            errMsg = tagKeyInvalid;
+        }
+        if (item.Value.length > 256) {
+            errMsg = tagValueInvalid;
+        }
+    });
+    if (errMsg) {
+        log.debug(errMsg,
+            { method: 'bucketPutTagging',
+                error: errors.InvalidTag });
+        const error = errors.InvalidTag.customizeDescription(
+            errMsg);
+        return cb(error);
+    }
+    return cb(null, { tags, bucket });
+}
+
+/**
+ * Bucket Put Tagging - Create or update bucket Tagging
+ * @param {AuthInfo} authInfo - Instance of AuthInfo class with requester's info
+ * @param {object} request - http request object
+ * @param {object} log - Werelogs logger
+ * @param {function} callback - callback to server
+ * @return {undefined}
+ */
+function bucketPutTagging(authInfo, request, log, callback) {
+    log.debug('processing request', { method: 'bucketPutTagging' });
+
+    const parser = new Parser({ explicitArray: false });
+    const bucketName = request.bucketName;
+    const metadataValParams = {
+        authInfo,
+        bucketName,
+        requestType: 'bucketPutTagging',
+    };
+    return waterfall([
+        next => _parseXML(request, log, next, parser),
+        (result, next) => metadataValidateBucket(metadataValParams, log,
+            (err, bucket) => next(err, { result, bucket })),
+        ({ result, bucket }, next) => checkTags(result.Tagging.TagSet.Tag, bucket, log, next),
+        ({ tags, bucket }, next) => {
+            bucket.setTags(tags);
```

Comment:
Random question, the only place I found this function was in ObjectMD (in Arsenal), but the function overwrite all the tags with the previously-set tags. In your tests, can you try to put two tags (one by one) and check, after the two PutBucketTagging, that both tags are in the bucket MD?

Reply:
Didn't see the api note: "When this operation sets the tags for a bucket, it will overwrite any current tags the bucket already has. You cannot use this operation to add tags to an existing list of tags.", so all good